### PR TITLE
dawn4py tests refactoring

### LINKED ISSUE #764: Work on danw4py integration tets
These are currently tests from SIR -> Generated code. They should be smaller and do SIR -> IIR

also consider reworking the IIR -> codegen part into into codegen tests

### PR DESCRIPTION
**Pull request by Stagno**
## Technical Description

- Splits the dawn4py integration tests into unit tests
  - dawn4py to serialized SIR (against reference) in `unit-test/dawn4py`
  - IIR to generated code (against reference) in `unit-test/dawn/CodeGen`
- Adds, to the CMake in `unit-test/dawn/CodeGen`, targets to regenerate input IIRs (from the dawn4py samples) and to update the references (by running `dawn-codegen` on the input IIRs).
- Adds CMake to update references of `unit-test/dawn4py`
- Fixes a few small bugs:
    - serialization of a reduction expression
    - backend string parsing
    - dawn-codegen input format deduction making dump to the same as dawn-codegen

### Resolves / Enhances

Resolves #764.

### CODE REVIEW

**Comment by Stagno** [comment]
launch jenkins

**Comment by Stagno** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

File: dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import dawn4py
+from dawn4py.serialization import SIR
+from dawn4py.serialization import utils as sir_utils
+from google.protobuf.json_format import MessageToJson
+
+stencil_name = "copy_stencil"
+output_file = f"{stencil_name}.cpp"
+
+if __name__ == "__main__":
+    interval = sir_utils.make_interval(SIR.Interval.Start, SIR.Interval.End, 0, 0)
```
Comment (twicki) [inline dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py]: Is there a reason why the main code is both setup and then compilation? Its would intuitively look nicer if we have a function that sets the SIR and than have the main just call 
```
setup()
compile()
rename()
```
Thoughts?

File: dawn/test/unit-test/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
 add_subdirectory(dawn)
 add_subdirectory(dawn-c)
+add_subdirectory(dawn4py)
```
[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/test/unit-test/CMakeLists.txt]: they were integration tests when they tested dawn4py -> generated code... now the ones under unit-test/dawn4py only test the dawn4py bindings generating the expected SIR

File: dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py (hunk 1 above)
Comment (Stagno) [reply dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py]: I see the benefit of separating the setup part from the rest. However i would keep these 2 lines together
```
dawn4py.compile(sir, backend="cuda", serialize_iir=True, output_file=output_file)
os.rename(stencil_name + ".0.iir", stencil_name + ".iir")
```
as the "module" which would make most sense to me here is `sir_to_iir()`, but I have no way to specify to dawn4py that I don't want the `.0` extension.

**Comment by Stagno** [comment]
@twicki @jdahm how should we proceed with this PR?

**Comment by jdahm** [comment]
@Stagno If @twicki accepts this PR, then let's fix the conflicts and merge it into master. This will likely cause many conflicts with the DawnCompiler PR, but it's closer to master than the other.

**Comment by jdahm** [comment]
However, maybe make it easy to regenerate the reference files (or tell me how it's done), because we'll have to do that again on the DawnCompiler PR right after merging this.

**Comment by Stagno** [comment]
@jdahm to update the references of tests in `unit-test/dawn/CodeGen` you can use the new `UpdateCodeGenReferences` cmake target, which will use the `dawn-codegen` executable, see
https://github.com/MeteoSwiss-APN/dawn/blob/3f60d9e7c7eb9a80916815271b9e7082e50580c5/dawn/test/unit-test/dawn/CodeGen/reference/CMakeLists.txt
To update the ones in `unit-test/dawn4py` you'd need to first run the tests and then use the target `UpdateDawn4PyUnittestsReferences` which will simply copy the `.sir` files from the build folder

**Comment by Stagno** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

File: dawn/test/unit-test/dawn4py/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (jdahm) [inline dawn/test/unit-test/dawn4py/CMakeLists.txt:1]: We created dawn4py-tests to avoid a problem with dawn4py. Should these go in there?

**Comment by jdahm** [comment]
@Stagno This is next on my list to finish. I'll do this tomorrow.

File: dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    # create the out = in[i+1] statement
+    body_ast = sir_utils.make_ast(
+        [
+            sir_utils.make_assignment_stmt(
+                sir_utils.make_field_access_expr("out", [0, 0, 0]),
+                sir_utils.make_field_access_expr("in", [1, 0, 0]),
+                "=",
+            )
+        ]
+    )
```
Comment (jdahm) [inline dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py:40]: @Stagno this looks like a really long line. If you can, please turn on the black formatter in your editor.

Comment (jdahm) [reply dawn/test/unit-test/dawn/CodeGen/samples/copy_stencil.py:40]: Sorry about that, my editor put this comment on the wrong line. However, there are long lines in here.

**Comment by jdahm** [comment]
@Stagno this is a bigger task than I thought. Will split up into a few PRs and add you as a reviewer.

**Comment by jdahm** [comment]
Closing temporarily and tracking in https://github.com/MeteoSwiss-APN/dawn/issues/949

**Pull request closed without merging**